File: benchmarks/tasks/bst_insert/bst_insert.py
# SPDX-License-Identifier: MIT-0

def insert(tree, v):
    if tree is None:
        return (None, v, None)
    left, val, right = tree
    if v < val:
        return (insert(left, v), val, right)
    elif v > val:
        return (left, val, insert(right, v))
    return tree

def inorder(tree):
    if tree is None:
        return []
    left, val, right = tree
    return inorder(left) + [val] + inorder(right)

tree = None
for v in [5, 3, 7, 1, 4]:
    tree = insert(tree, v)

print(" ".join(str(x) for x in inorder(tree)))

File: benchmarks/tasks/bst_insert/bst_insert.hs
-- SPDX-License-Identifier: MIT-0
module Main where

data BST = Empty | Node Int BST BST

insert :: Int -> BST -> BST
insert x Empty = Node x Empty Empty
insert x (Node v left right)
  | x < v    = Node v (insert x left) right
  | x > v    = Node v left (insert x right)
  | otherwise = Node v left right

inorder :: BST -> [Int]
inorder Empty            = []
inorder (Node v left right) = inorder left ++ [v] ++ inorder right

main :: IO ()
main = putStrLn (unwords (map show (inorder tree)))
  where
    tree = foldr insert Empty [4, 1, 7, 3, 5]
